### ISSUE TITLE
add the assistants API

### ISSUE THREAD

**Issue by restlessronin** [restlessronin/openai_ex#55]
### Describe the feature or improvement you're requesting

https://platform.openai.com/docs/assistants/overview

### Additional context

- [x] Implement Assistants
- [x] Implement Threads
- [x] Implement Messages
- [x] Implement Runs
- [x] Add Beta documentation to User Guide
- [x] Check all existing notebooks are still working.

**Comment by TomBers** [comment]
@restlessronin I was just about to open an issue to mention it.  I guess this means you are already on it!

**Comment by restlessronin** [comment]
@TomBers yeah, I started work on the new API. Unfortunately there's no easy way to divide the work and it's a small enough task that it's not worth spending much time figuring out a division of labour 😉

There may be another independent task though. It seems like there have been some changes in the chat completion API at the same time (deprecated 'function' keywords for equivalent 'tools' keywords, an unclear example of image-input, and possibly others). If you could take a crack at another PR that makes those changes, that would save me having to do it. Since not too many people are using v 0.3.0 that implemented 'function' I think it can be safely dropped in favor of 'tools' in v 0.4.  WDYT?

**Comment by TomBers** [comment]
@restlessronin I will look into it, and if I can get my head around the changes will open a PR.  Good luck with the Assistants work!

**Comment by TomBers** [comment]
@restlessronin I am going to go step by step, so I have opened a PR with the update from function -> tools (updated in the doc as well), Happy to take out the legacy function_call (although from my understanding the whole chat completion is depreciated? Alot is going on).  I am just looking at the image-input.

**Comment by TomBers** [comment]
@restlessronin I was looking at the image input, I came across this example: 

```
client.chat.completions.create(
    model="gpt-4-vision-preview",
    messages=[
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "What’s in this image?"},
                {
                    "type": "image_url",
                    "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                },
            ],
        }
    ])```

https://platform.openai.com/docs/api-reference/chat/create

To my mind, it seems it is just passing `type` and  `url`.  Unless I am missing something, I am not sure what needs to be done.

**Comment by restlessronin** [comment]
From what I recall, there was a different example when I looked (perhaps the docs were still changing rapidly). This new example is fairly clear and supporting it doesn't seem to require anything except documentation of the use case.

**Comment by TomBers** [comment]
@restlessronin Thanks for merging the PR.  I did open a 2nd to document the image input.  Good talking to you, and have a good weekend.

**Comment by TomBers** [comment]
@restlessronin Unfortunately I do not have too much time this weekend, but I did a few test runs on the Livebooks, I hope it helps.  

The only failure is ironically with the function calling:

<img width="854" alt="image" src="https://github.com/restlessronin/openai_ex/assets/1389427/6d5deb1d-41ca-46dd-bcd7-5d697577bd1e">

It is not immediately obvious to me what the error is indicating.

I had a few 'does this look right moments' on the Completion Bot Livebook.

<img width="873" alt="image" src="https://github.com/restlessronin/openai_ex/assets/1389427/b1d28472-72fe-41fc-836b-c3cb2fd11bac">


I was expecting some sort of GUI, likewise the Streaming.

<img width="873" alt="image" src="https://github.com/restlessronin/openai_ex/assets/1389427/7e4cabb0-0785-426c-927e-a64a78204460">


I hope that helps.

**Comment by restlessronin** [comment]
@TomBers thanks for doing the checks. I have corrected and updated the existing documentation and published v0.4.0 to hexpm.

I need to complete the documentation for the Beta Assistants API to include in the User Guide. I hope to get that done tomorrow. Once that is done, I will close this issue.

In the meantime I will announce the new release on the forum.

**Comment by restlessronin** [comment]
I have just published v0.4.1 with

1. A first cut at adding all the new beta api calls to the user guide.
2. DALL-E-3 support in the Image endpoint (which I somehow missed in my first scan for API changes)
3. changes to the FQN for some modules (although I did not bump up the major version)

I am closing this issue as complete.

if there’s anything missing or not working as it should, please file an issue (or better yet, a PR  😀 Likewise for any suggestions for improvement.
